### PR TITLE
new(tests): TSTORE: ensure transient storage is cleared after transactions

### PR DESCRIPTION
**Pull request by jochem-brouwer**
## 🗒️ Description
This PR introduces 2 tests, which test that the transient storage is cleared after transactions. EthereumJS did clear storage after "normal" transactions, but did not clear the transient storage after a transaction created a contract. If the deployed contract address was then called, then the transient storage was not empty. See: https://github.com/ethereumjs/ethereumjs-monorepo/pull/3625. This was not caught by these tests, so I have added two tests, one which does the create-tx and then call-contract, and the other calls the same contract twice.

The tests are very repetitive, so I would love some suggestions how to make those more compact (we could also move these two tests into one single test, where we use 3 transactions instead of two, writing to different storage slots and different transient storage slots).

The good/bad cases of these tests have been tested using EthereumJS (with and without the bug) using our new t8ntool :smile: 

## 🔗 Related Issues
Test produced by a bug in EthereumJS: https://github.com/ethereumjs/ethereumjs-monorepo/pull/3625

## ✅ Checklist
- [ ] All: Set appropriate labels for the changes.
- [ ] All: Considered squashing commits to improve commit history.
- [ ] All: Added an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [ ] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [ ] Tests: All converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been added to [converted-ethereum-tests.txt](/ethereum/execution-spec-tests/blob/main/converted-ethereum-tests.txt).
- [ ] Tests: A PR with removal of converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been opened.
- [ ] Tests: Included the type and version of evm t8n tool used to locally execute test cases:  e.g., ref with commit hash or geth 1.13.1-stable-3f40e65.
- [ ] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://ethereum.github.io/execution-spec-tests/main/tests/) are correctly formatted.

### CODE REVIEW

File: tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py
Comment (marioevz) [inline tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py:29]:
```suggestion
@pytest.mark.valid_from("Cancun")
@pytest.mark.with_all_evm_code_types
def test_tstore_clear_after_deployment_tx(
    blockchain_test: BlockchainTestFiller,
    pre: Alloc,
    evm_code_type: EVMCodeType,
):
```

I feel like we could expand this test a little bit to support EOF by simply adding this marker, and then in the test doing something like:

```python
if evm_code_type == EVMCodeType.EOF_V1:
  code = Container.Init(deploy_container=Container.Code(Op.SSTORE(1, Op.TLOAD(1)) + Op.STOP), initcode_prefix=Op.TSTORE(1, 1))
else:
  code = Initcode(deploy_code=deploy_code, initcode_prefix=init_code)
```
As of now it will not run in EOF until the test is filled in a fork where EOF is active.

**Review by marioevz — COMMENTED** [review]
Looks good! Just one comment, thanks!

File: tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py
Comment (jochem-brouwer) [reply tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py]: Have added!

File: tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py
Comment (holgerd77) [inline tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py:54]: Doesn’t this line invalidate the previous conditional association?

File: tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py
Comment (jochem-brouwer) [reply tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py:67]: Yes, this sends a transaction without any calldata (empty calldata) to the contract we just deployed.

File: tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py
Comment (jochem-brouwer) [reply tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py:54]: Woops, it does!! I should resolve this

Comment (jochem-brouwer) [reply tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py]: Should be resolved

File: tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py
Comment (jochem-brouwer) [reply tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py:67]: ("invocation" would mean the same as "call a contract" here, I might have messed up some grammar/spelling/definitions or whatever this would be called)

File: tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py
Comment (marioevz) [inline tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py:100]:
```suggestion
    code = Op.SSTORE(1, Op.TLOAD(1)) + Op.TSTORE(1, 1) + Op.STOP
    account = pre.deploy_contract(code)
```

The logic here shouldn't be necessary for this case, pre.deploy_contract changes its behavior when we use pytest.mark.with_all_evm_code_types to read fixture evm_code_type and automatically wrap the deployed contract within an EOF container (it also should automatically add the STOP).

Might need to remove Bytecode import after this change.

File: tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py
Comment (marioevz) [inline tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py:46]:
```suggestion
    deploy_code = Op.SSTORE(1, Op.TLOAD(1)) + Op.STOP
```

**Review by marioevz — APPROVED** [review]
LGTM, thanks!

File: tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py
Comment (jochem-brouwer) [reply tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py:46]: I was about to comment "we don't need `Op.STOP`", but this to make the EOF container valid - right?

File: tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py
Comment (jochem-brouwer) [reply tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py:100]: Ok, so I can remove `Op.STOP` and then just remove the `evm_code_type` switch logic?

I think I actually copied this logic from another test, so this logic (not automatically switching to EOF) is somewhere else (unless that is necessary for that test). I actually took this from the `test_code_set_txs.py` (EIP 7702), but looking there again I think that logic needs to be there because of the contract creations? :thinking: 

Would the logic still be necessary in `test_tstore_clear_after_deployment_tx` since that deploys a contract?

File: tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py
Comment (marioevz) [reply tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py:46]: Yes that is correct, only because of EOF.

File: tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py
Comment (marioevz) [reply tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py:100]: > Ok, so I can remove `Op.STOP` and then just remove the `evm_code_type` switch logic?

Yes, `Op.STOP` is optional because `pre.deploy_contract` detects that the code is not `Container`, it's `Bytecode`, so before automatically wrapping, it checks for the `terminating` flag of the bytecode, and if it's false, it appends the `Op.STOP`.

> I think I actually copied this logic from another test, so this logic (not automatically switching to EOF) is somewhere else (unless that is necessary for that test). I actually took this from the `test_code_set_txs.py` (EIP 7702), but looking there again I think that logic needs to be there because of the contract creations? 🤔
> 
> Would the logic still be necessary in `test_tstore_clear_after_deployment_tx` since that deploys a contract?

Yes I think that's the reason,  if we are dealing with initcode, it needs the switch because we don't handle the `evm_code_type` flag automatically inside the `Initcode` class as we do with `pre.deploy_contract`.

Comment (jochem-brouwer) [reply tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py]: Have removed the EOF code switch logic in this specific test.

Comment (jochem-brouwer) [reply tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py]: I will merge once the CI passes :)

Comment (jochem-brouwer) [reply tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py]: Side-question, why does this PR stay approved even since I just pushed a commit? Shouldn't this require a re-review? I could now put arbitrary code here and merge it to `main`? :thinking:

Comment (jochem-brouwer) [reply tests/cancun/eip1153_tstore/test_tstorage_clear_after_tx.py]: Oh wait, nvm, I cannot merge it myself

**Pull request merged**
